gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -276,39 +259,46 @@ hybridElimination(const HybridGaussianFactorGraph &factors,
       frontalKeys, continuousSeparator, discreteSeparator, conditionals);
 
-  // If there are no more continuous parents, then we should create a
-  // DiscreteFactor here, with the error for each discrete choice.
   if (continuousSeparator.empty()) {
-    auto factorProb =
-        [&](const GaussianMixtureFactor::FactorAndConstant &factor_z) {
-          // This is the probability q(μ) at the MLE point.
-          // factor_z.factor is a factor without keys,
-          // just containing the residual.
-          return exp(-factor_z.error(VectorValues()));
-        };
-
-    const DecisionTree<Key, double> fdt(newFactors, factorProb);
-    // // Normalize the values of decision tree to be valid probabilities
-    // double sum = 0.0;
-    // auto visitor = [&](double y) { sum += y; };
-    // fdt.visit(visitor);
-    // // Check if sum is 0, and update accordingly.
-    // if (sum == 0) {
-    //   sum = 1.0;
-    // }
-    // fdt = DecisionTree<Key, double>(fdt,
-    //                                 [sum](const double &x) { return x / sum;
-    //                                 });
-    const auto discreteFactor =
-        boost::make_shared<DecisionTreeFactor>(discreteSeparator, fdt);
+    // If there are no more continuous parents, then we create a
+    // DiscreteFactor here, with the error for each discrete choice.
+
+    // Integrate the probability mass in the last continuous conditional using
+    // the unnormalized probability q(μ;m) = exp(-error(μ;m)) at the mean.
+    //   discrete_probability = exp(-error(μ;m)) * sqrt(det(2π Σ_m))
+    auto probability = [&](const Result &pair) -> double {
+      static const VectorValues kEmpty;
+      // If the factor is not null, it has no keys, just contains the residual.
+      const auto &factor = pair.second;
+      if (!factor) return 1.0;  // TODO(dellaert): not loving this.
+      return exp(-factor->error(kEmpty)) / pair.first->normalizationConstant();
+    };
 
+    DecisionTree<Key, double> probabilities(eliminationResults, probability);
     return {boost::make_shared<HybridConditional>(gaussianMixture),
-            discreteFactor};
+            boost::make_shared<DecisionTreeFactor>(discreteSeparator,
+                                                   probabilities)};
   } else {
-    // Create a resulting GaussianMixtureFactor on the separator.
+    // Otherwise, we create a resulting GaussianMixtureFactor on the separator,
+    // taking care to correct for conditional constant.
+
+    // Correct for the normalization constant used up by the conditional
+    auto correct = [&](const Result &pair) -> GaussianFactor::shared_ptr {
+      const auto &factor = pair.second;
+      if (!factor) return factor;  // TODO(dellaert): not loving this.
+      auto hf = boost::dynamic_pointer_cast<HessianFactor>(factor);
+      if (!hf) throw std::runtime_error("Expected HessianFactor!");
+      hf->constantTerm() += 2.0 * pair.first->logNormalizationConstant();
```

Comment:
@varunagrawal @ProfFan this is the correction that was missing.

Reply:
Ah! I think this is the same method in Murphy02thesis, using the HessianFactor (which is in the canonical form of a Gaussian).

I regret totally missing this...